MSC2790: Modal widgets (acquiring user input from a widget) (#2790)
matrix-org/matrix-doc: proposals/2790-modal-widgets.md

the request with an empty response object.

Once the client receives a `close_modal` request from the modal widget, the client MUST terminate that
widget's session and close the modal. It should respond with an empty response object first, however
by nature of severing the connection the modal widget should not have an issue with never receiving a
reply. The client must reject any subsequent `close_modal` requests from the modal widget.

Between a successful `open_modal` and `close_modal` request, the source widget can assume that the
user is being prompted as expected. This is typically useful for the source widget to prevent accidental
interaction from the user while it is acquiring more information.

Note: the `m.close` button ID does *not* automatically close the widget. The modal widget is responsible
for handling the button click and requesting a close if required. If the user dismissed the modal
widget without using the provided buttons (such as a back button or clicking on a backdrop), the client
immediately terminates the modal widget's session without informing it and sends a `close_modal`

> request to the source widget with an `"m.exited": true` parameter, like so:

Author: this is a potential trap. Maybe clients should automatically close the modal after some seconds? can a widget offer zero close buttons?

the client's built-in styles where possible. This is done through a new kind of widget called
"modal widgets" and have some additional restrictions on them.

As of writing, the best resource for reading on the widget specification is the work-in-progress
PR: https://github.com/matrix-org/matrix-doc/pull/2764

## Proposal

In addition to the existing room and account widgets, a new widget type, "modal", is introduced.
Modal widgets can be spawned by room and account widgets only - it is not intended to allow modal
widgets to stack. Modal widgets are not persisted in any particular location.

To help with permissions, modal widgets can only be spawned if the *source widget* (the widget
which opened the modal widget) has the `m.modals` capability. As with all capabilities, users should
be prompted to approve the widget's use of this new capability.

Modal widgets SHOULD be rendered in a dialog-like component within the client to get the user's
attention. The dialog should have a title, possibly subtitle, the modal widget, and any buttons
requested by the source widget. The dialog-like component is also called the widget's frame in

> this proposal.

Author: it probably should, but I wouldn't make this a requirement. Will incorporate.

A modal widget takes a very similar shape to the common widget properties definition:

```json
{
  "type": "m.custom",
  "url": "https://example.org/modal_widget.html?user_id=$matrix_user_id",
  "name": "What is your name?",
  "data": {
    "custom-key": "This is a custom key"
  },
  "waitForIframeLoad": true,
  "buttons": [
    {"id": "com.example.save", "label": "Submit", "kind": "m.primary"},

>     {"id": "m.close", "label": "Close", "kind": "m.link"}

Reviewer: The example image gives this a text of "Cancel" - should they match?

Author: fixed manually

  ]
}
```

> Most notably, the `creatorUserId` is not supported on modal widgets and is implied to be the currently

Reviewer: If the definition specifies it anyway, is it ignored or does that cause an error?

Author: ignored - it's like everything else in Matrix.

Reviewer: it's not obvious to me that it should be ignored. It's certainly a valid decision, but one worth making explicit.

Author: I'll soften the language here, but it's a problem for the whole spec and not this MSC in particular.

active user ID. All the other properties in common with the standard widget definition keep their
meanings and requirements as specified.

The `buttons` array is only valid on modal widgets and is optional. When not supplied or invalid, its
value is implied to be an empty array (no buttons). Buttons have 3 properties to them, all required:

* `id` - Using [MSC2758](https://github.com/matrix-org/matrix-doc/pull/2758), an identifier to indicate
  which button was clicked by the user to the modal widget. Only `m.close` is specified by this MSC,
  allowing clients to render any applicable close buttons on their UI.

> * `label` - A human-readable string for the text of the button.

Reviewer: What happens if it's not a string?

Author: clients handle it appropriately? not sure we need to put data validation on every single field - this is a larger spec concern.

Reviewer: as a client I could plausibly
 * ignore the button
 * show the button but with a placeholder label
 * try to do something fancy

maybe it's fine if clients all do different things, but this sort of implementation-specific behaviour is a great path to widgets that only work with specific clients. 

> not sure we need to put data validation on every single field

I kinda disagree here. We've made a lot of problems for ourselves and the Matrix protocol by failing to specify what counts as "valid", and what should happen when something is "not valid". It'd be nice to try to avoid making it worse.

Author: The widget API has an error reporting structure to its API - we just need to formalize its use for invalid data. I still think that extends beyond this MSC.

Reviewer: right, but below you've specced that clients should *ignore* invalid buttons, not report errors?

yes, the problem that we have spectacularly failed to specify what constitutes valid data and what to do with invalid data extends beyond this MSC. I can't see that is a justification for piling on even more debt?

Author: I'm writing the MSC as it applies to widgets while having this conversation, so it's not necessarily piling on.

If the MSC already says that invalid buttons are ignored, then I'm not sure what this thread is about?

Reviewer: this thread is about what counts as valid. In particular, does `label` not being a string make it invalid? If the answer is simply "yes", then just make it explicit and I'll be happy :)

Author: https://github.com/matrix-org/matrix-doc/pull/2813 for the general validation, will clarify this here.

> * `label` - A human-readable string for the text of the button.

Author: the entire spec ignores i18n, not just widgets 😛

good point though - widgets are a bit special and have a higher chance at being in the user's language.

* `kind` - Using MSC2758 again, the kind of button the client should present the button as. Clients
  which do not understand the `kind` or do not have styling rules for the `kind` should render the

>   button as a `m.primary` button. This MSC defines the following common types:

Reviewer: Primary seems like a surprising choice. Not secondary?

Author: what would the argument be for secondary?

Author: ah, fair point - will fix.

Reviewer: @turt2live this does not appear to have been updated?

Reviewer: Sorry, was reading an outdated copy!

  * `m.primary` - A button which is styled as the primary action. Usually this will mean a background
    of the client's accent colour.

>   * `m.secondary` - A button which is styled similar to the primary button but denotes secondary action.

Author: it could be - there's not really anything against this. The guidelines are just "not the primary button" style. I'll try to reword this.

  * `m.warning` - A button which is styled to warn the user of potential harm in the action about to
    be taken. Usually yellow or orange backgrounds are used.
  * `m.danger` - A button which is styled to warn the user of imminent risk of severe consequences,
    such as deleting an object. Usually these are red in background colour.
  * `m.link` - A button which looks like a link but has sizing to have it match other buttons.

The client is not required to have styling rules for all button `kind`s, though is encouraged to
support at least the ones listed in this MSC. Clients should ignore any buttons missing any of the
3 required fields.

An example rendering of the above example widget would be as follows:

![2790-widget-modal-example.png](./images/2790-widget-modal-example.png)

In this example, the title is the `name` of the widget (a subtitle could be shown from the `title`
attribute under `data`, if present). The iframe in the middle is simply the modal widget itself,

> rendered just like any other widget would (fill in the `url` template, render). The buttons at

Reviewer: ```suggestion
rendered just like any other widget would be (fill in the `url` template, render). The buttons at
```

Author: applied manually

the bottom are specified by the `buttons` array.

The example image has a close button in the top right too - this is an example of the client
rendering the `m.close` ID'd button in custom UI. When clicked, this would behave exactly like

> the "Close" button on the bottom.

Reviewer: ```suggestion
the "Cancel" button on the bottom.
```

Author: applied manually

Clients should not ask for the user's permission to open the modal because the capability will have
already been approved. A widget which receives a successful response to `open_modal` can expect
the user to be currently prompted with the modal (this matters more later).

The modal widget dialog is like any other widget with a few added actions. The modal widget gets
assigned a widget ID that is generated by the client, and the regular widget API steps are set up.
Due to the nature of modal widgets and their use case, modal widgets MUST support the widget API.
Similarly, clients MUST handle the widget API when interacting with a modal widget. If the source
widget's session ends before the modal widget is closed, the client MUST close the modal widget
and terminate its session.

This MSC does not define any restrictions for the number of modal widgets that can be active at
any one time, and does not define behaviour for how to handle multiple modal widgets. The client
is expected to remember which widget opened which modal, however - more on this later in the
proposal.

After the modal widget's session is established (capabilities exchanged), the client MUST send

> a `toWidget` API request informing the widget of its configuration. The source widget is able to

Reviewer 2: ```suggestion
a `toWidget` API request informing the modal widget of its configuration. The source widget is able to
```

pass data through to the modal widget by making use of custom keys in the widget `data`. This request
looks like the following:

```json
{
  "api": "toWidget",
  "action": "widget_config",
  "requestId": "AAABBB",
  "widgetId": "CCCDDD",
  "data": {
    "type": "m.custom",
    "url": "https://example.org/modal_widget.html?user_id=$matrix_user_id",
    "name": "What is your name?",

>     "data": {

Author: This is actually to keep as much of the widget schema as possible. A normal widget has a `data` at the root of it where custom keys are placed - it would be good to be able to reuse the same here for templating and such.

The `data` for a payload in a widget request is also annoying as a conflict (particularly when browsers also introduce a *third* `data` off postMessage events), however that is best left to another MSC to resolve.

>     "data": {

Reviewer 3: ah right, fair enough

modal widget to get the user's event details using an `open_modal` request:

```json
{
  "api": "fromWidget",
  "action": "open_modal",
  "requestId": "AAABBB",
  "widgetId": "CCCDDD",
  "data": {
    "type": "com.example.calendar.new_event",
    "url": "https://example.org/event_details.html?user_id=$matrix_user_id",
    "name": "New event details",
    "data": {
      "calendarId": 53,
      "startDate": "2020-09-22"
    },
    "buttons": [
      {"id": "com.example.save", "label": "Add Event", "kind": "m.primary"},
      {"id": "m.close", "label": "Cancel", "kind": "m.link"}
    ]
  }
}
```

In this example, the client has already approved the widget's capability to open modals and doesn't
have an issue with the widget definition, so it responds to the request and shows the widget to the
user. Note that in this example the widget has a custom `type` to allow the client to natively
render the widget instead of using an iframe if it so desired.

After the modal widget's session has been initialized and the widget config sent over to it, the
user enters information into the widget. Once complete, they click the "Add Event" button at the

> bottom of the dialog. The modal widget sees this and packages up the user's information into a

Reviewer 2: > The modal widget sees this ...

How does the modal widget notice that the user has clicked on a button that was rendered by the client?

Reviewer 4: https://github.com/matrix-org/matrix-doc/pull/2790#issuecomment-705598247 - the Client sends a `button_pressed` event

`close_modal` request:

```json
{
  "api": "fromWidget",
  "action": "close_modal",
  "requestId": "AAABBB",
  "widgetId": "CCCDDD",
  "data": {
    "calendarId": 53,
    "date": "2020-12-31",
    "startTime": "00:00",
    "durationHours": 24,
    "title": "New Year's Eve party",
    "description": "Something witty would go here",
    "invitees": [
      "@alice:example.org",
      "mailto:bob@example.org"
    ]
  }
}
```

> The client acknowledges this, terminates the modal widget, and proxies a similar `close_modal`

Reviewer 2: > ... and proxies a similar ...

If it's just "similar", what are the differences?

request to the original calendar widget. The calendar widget then processes this information and
adds the event to the calendar.

Note that the user interaction flow doesn't have to be this way: the modal widget could just as
easily use the same backend as the calendar widget and add the event directly. It would simply send

> and empty `data` objects through, and the calendar widget would likely ignore the request anyways.

Reviewer 2: ```suggestion
an empty `data` object through, and the calendar widget would likely ignore the request anyways.
```

I think?

A potential reason for using this data channel is because the calendar widget may have already gone
through all the hoops to validate the user's identity and authentication, and while the modal can
do all those same steps again it could be disruptive to the user's experience.

## Potential issues

> Modal widgets don't have a way to disable buttons inthe dialog with this proposal. This is currently

Reviewer 2: ```suggestion
Modal widgets don't have a way to disable buttons in the dialog with this proposal. This is currently
```

> Modal widgets don't have a way to disable buttons inthe dialog with this proposal. This is currently

Author: duplicate https://github.com/matrix-org/matrix-doc/pull/2790/files#r510965254

a limitation and not planned to be included by this MSC. Likewise are other common button kinds

> like "secondary", "info", "light", and "dark".

Reviewer 2: But the MSC does define a "secondary"?

```suggestion
like "info", "light", and "dark".
```

## Alternatives

A modal widget could instead be the whole dialog (title, buttons - everything), however this has a
number of concerns relating to styling the buttons for every single client possible. The widget API
would also need to be extended to pass through theme information, which it may very well want to do

> anyways, but would still be added complexity for a simple form-style widget.

Author: > This is true for any other type of widget though?

Yes, however modal widgets won't always be a modal dialog. On mobile clients for instance, it might be desirable to show a card or pop-in effect rather than a true dialog. In these cases, the widget would have to consider the brand, theme, and platform of the app which leads to infinite amounts of potential CSS. Instead, we pull the buttons and some other basic information out so the client can render it as natively as possible.

> Does it not have all the same problems anyway for displaying the URL part of the modal widget?

which problems are you thinking of?

> anyways, but would still be added complexity for a simple form-style widget.

Reviewer 3: Taking the example of the calendar add on mobile, would you not still have to have infinite amounts of CSS for the event details to match the brand / theme / etc? ie. you'll get a native, "New Event Details", native buttons and then the time / place / description in a white box in Times New Roman.

> anyways, but would still be added complexity for a simple form-style widget.

Author: Not necessarily, if taking the perspective of a cross-platform designer. A widget is typically displayed in a context where the user is aware that it's not native UI, so variances (or even completely different themes) are acceptable. A modal widget would want to provide as much native UI as possible though to give a better user experience within the app. It's much less of a concern for web apps as the CSS becomes super easy to do, but for iOS, Android, and other mobile devices there's not really a good CSS library that makes things appear as other elements.

It's a bit convoluted, but it also plays into a theory that clients can ignore the `url` and just natively render the widget (form) if they understand enough of the detail.

> anyways, but would still be added complexity for a simple form-style widget.

Reviewer 5: If helpful for extra context; the intent here is for modal widgets to default to native interactions for the call to actions on the platform they're being viewed on.

> anyways, but would still be added complexity for a simple form-style widget.

Reviewer 3: Ah, I think the 'type' that allows clients to know what the content should be and render it natively is what I was missing here. The only mention I can see of this is in the real-world example part?

> anyways, but would still be added complexity for a simple form-style widget.

Author: @dbkr I've added a bit of clarifying text to the end of this paragraph. Let me know if it makes more (or less) sense.

widget to indicate as such. When terminating the modal widget this way, the client sends an `m.exited`
field to indicate that the user/client has cancelled the user input. An example request is:

```json
{
  "api": "toWidget",
  "action": "close_modal",
  "requestId": "AAABBB",
  "widgetId": "CCCDDD",
  "data": {
    "m.exited": true
  }
}
```

An example for when the client would want to terminate the widget this way is when the user clicks on
the background around the dialog, thus closing the dialog, or when the client believes the widget is
acting in bad faith and refusing to close (eg: not closing within 5 seconds of clicking the close button).

### Disabling buttons

Modal widgets might want to disable some of the buttons until the user has entered enough information
to cause them to become enabled. They can do so with the following `fromWidget` API actions. Buttons
are always enabled by default.

```json
{
  "api": "fromWidget",
  "action": "set_button_enabled",
  "requestId": "AAABBB",
  "widgetId": "CCCDDD",
  "data": {
    "button": "com.example.save",
    "enabled": false
  }
}
```

The response to which is an empty `response` object:

```json
{

>   "api": "fromWidget",

Reviewer: toWidget?

Author: nope - the response to a widget API action is always a copy of the request plus a `response` field.

Reviewer: how confusing. Fair enough.

  "action": "close_modal",
  "requestId": "AAABBB",
  "widgetId": "CCCDDD",
  "data": {
    "calendarId": 53,
    "date": "2020-12-31",
    "startTime": "00:00",
    "durationHours": 24,
    "title": "New Year's Eve party",
    "description": "Something witty would go here",
    "invitees": [
      "@alice:example.org",
      "mailto:bob@example.org"
    ]
  }
}
```

The client acknowledges this, terminates the modal widget, and proxies a similar `close_modal`
request to the original calendar widget. The calendar widget then processes this information and
adds the event to the calendar.

Note that the user interaction flow doesn't have to be this way: the modal widget could just as
easily use the same backend as the calendar widget and add the event directly. It would simply send
an empty `data` object through, and the calendar widget would likely ignore the request anyways.
A potential reason for using this data channel is because the calendar widget may have already gone
through all the hoops to validate the user's identity and authentication, and while the modal can
do all those same steps again it could be disruptive to the user's experience.

## Potential issues

> Modal widgets don't have a way to disable buttons in the dialog with this proposal. This is currently

Reviewer: Yes they do?

Reviewer: @turt2live can you update this?

Author: will do next time I'm in the area of this MSC. For now I've cancelled FCP because it looks like we instead might want to change how widgets work.
